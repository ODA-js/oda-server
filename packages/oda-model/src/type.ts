import { ElementMetaInfo, Internal, MetaData } from './element';
import {
  IModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import {
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  HashToArray,
  NamedArray,
  MapToArray,
  ArrayToMap,
} from './types';

import { merge } from 'lodash';
import * as inflected from 'inflected';
import { TypeFieldInput, TypeField, ITypeField } from './typefield';
import { EntityFieldInput } from './entityfield';
import { UIView } from './entitybase';

export interface IType extends IModelBase<TypeMetaInfo, TypeInput, TypeOutput> {
  readonly plural: string;
  readonly global: boolean;
  readonly titlePlural: string;
  readonly fields: Map<string, ITypeField>;
}

export interface TypeMetaInfo extends ElementMetaInfo {
  titlePlural: string;
  global: boolean;
  name: {
    plural: string;
    singular: string;
  };
  UI: UIView;
}

export interface TypeInternal extends ModelBaseInternal {
  fields: Map<string, ITypeField>;
}

export interface TypeInput extends ModelBaseInput<TypeMetaInfo> {
  plural?: string;
  titlePlural?: string;
  fields?: AsHash<TypeFieldInput> | NamedArray<TypeFieldInput>;
}

export interface TypeOutput extends ModelBaseOutput<TypeMetaInfo> {
  fields: NamedArray<TypeFieldInput>;
}

export const TypeDefaultMetaInfo = {
  name: {},
  UI: {
    listName: [],
    quickSearch: [],
    hidden: [],
    edit: [],
    show: [],
    list: [],
    embedded: [],
  },
};
export const typeDefaultInput = {
  metadata: TypeDefaultMetaInfo,
  exact: false,
};

export class Type
  extends ModelBase<TypeMetaInfo, TypeInput, TypeInternal, TypeOutput>
  implements IType {
  public get modelType(): MetaModelType {
    return 'input-type';
  }

  constructor(init: TypeInput) {
    super(merge({}, typeDefaultInput, init));
  }

  get plural(): string {
    return this.metadata.name.plural;
  }

  get titlePlural(): string {
    return this.metadata.titlePlural;
  }

  get global(): boolean {
    return this.metadata.global;
  }

  get fields(): Map<string, ITypeField> {
    return this[Internal].fields;
  }

  public updateWith(input: Nullable<TypeInput>) {
    super.updateWith(input);

    assignValue<TypeInternal, TypeInput, string>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => {
        src.name = value.trim();
        this.metadata.name.singular = src.name;
      },
      required: true,
    });

    assignValue<TypeMetaInfo, TypeInput, boolean>({
      src: this[MetaData],
      input,
      field: 'name',
      effect: (src, value) => (src.global = value),
      setDefault: src => (src.global = false),
    });

    assignValue<TypeMetaInfo, TypeInput, string>({
      src: this.metadata,
      input,
      field: 'name.plural',
      inputField: 'plural',
      effect: (src, value) => {
        src.name.plural = value.trim();
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
      setDefault: src => {
        src.name.plural = inflected.pluralize(this.name);
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
    });

    assignValue<TypeMetaInfo, TypeInput, string>({
      src: this.metadata,
      input,
      field: 'titlePlural',
      effect: (src, value) => (src.titlePlural = value.trim()),
      setDefault: src => (src.titlePlural = this.plural),
    });

    assignValue<
      TypeInternal,
      TypeInput,
      AsHash<TypeFieldInput> | NamedArray<TypeFieldInput>
    >({
      src: this[Internal],
      input,
      field: 'fields',
      effect: (src, value) => {
        const fields = ArrayToMap(
          Array.isArray(value) ? value : HashToArray(value),
          (fld, order) => {
            let field = new TypeField(
              merge({}, fld, {
                order,
                entity: this.name,
              } as Partial<EntityFieldInput>),
            );
            return field;
          },
          (obj: any, src: any) => obj.mergeWith(src.toObject()),
        );

        src.fields = fields;
      },
      setDefault: src => (src.fields = new Map()),
    });
  }

  public toObject(): TypeOutput {
    return merge({}, super.toObject(), {
      fields: MapToArray(this.fields, (name, value) => ({
        ...value.toObject(),
        name,
      })),
    } as Partial<TypeOutput>);
  }

  public mergeWith(payload: Nullable<TypeInput>) {
    super.mergeWith(payload);
  }
}
