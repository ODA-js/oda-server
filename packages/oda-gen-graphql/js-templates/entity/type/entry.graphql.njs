<#@ context 'entity' -#>
<#- if(entity.description){#>
# #{entity.description}
<# }#>

input Embed#{entity.name}Filter {
  or: [Embed#{entity.name}FilterItem]
  and: [Embed#{entity.name}FilterItem]
  some: #{entity.name}Filter
  none: #{entity.name}Filter
  every: #{entity.name}Filter
}

input Embed#{entity.name}FilterItem {
  some: #{entity.name}Filter
  none: #{entity.name}Filter
  every: #{entity.name}Filter
}

input #{entity.name}Filter {
  or: [#{entity.name}FilterItem]
  and: [#{entity.name}FilterItem]
<#-entity.filter.forEach((item, index)=>{#>
  #{item}
<#-})#>
}

input #{entity.name}ComplexFilter {
  or: [#{entity.name}ComplexFilter]
  and: [#{entity.name}ComplexFilter]
<#-entity.filter.forEach((item, index)=>{#>
  #{item}
<#-})#>
}

input #{entity.name}FilterItem {
<#-entity.filter.forEach((item, index)=>{#>
  #{item}
<#-})#>
}

input #{entity.name}FilterSubscriptionsItem {
<#-entity.filter.forEach((item, index)=>{#>
  #{item}
<#-})#>
}

input #{entity.name}FilterSubscriptions {
  or: [#{entity.name}FilterSubscriptions]
  and: [#{entity.name}FilterSubscriptions]
  mutation: WhereMutationKind
  node: #{entity.name}FilterSubscriptionsItem
  previous: #{entity.name}FilterSubscriptionsItem
  updatedFields: WhereListOfStrings
}

type #{entity.name} implements Node {
<#- entity.fields.forEach(field => { -#>
<# if(field.description){#>
  # #{field.description}
<#- }#>
  #{field.name}#{field.args}: #{field.type}
<#-})-#>

<#- entity.relations.forEach(rel => { -#>
<# if(rel.description){#>
  # #{rel.description}
<#- }#>
  <#-if(rel.single) {-#>

  #{rel.name}#{rel.args}: #{rel.type}

  <#-} else {-#>
  <#if(rel.derived){#>
  #{rel.name}(#{rel.args} ): #{rel.connectionName}
  <#} else {#>
  #{rel.name}(after: String, first: Int, before: String, last: Int, limit: Int, skip: Int, orderBy: [#{rel.entity}SortOrder], filter:#{rel.entity}ComplexFilter ): #{rel.connectionName}
  <#-}-#>
  <#-}-#>
<#-})#>
}


