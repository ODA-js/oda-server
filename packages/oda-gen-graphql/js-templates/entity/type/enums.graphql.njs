<#@ context 'entity' -#>
enum #{entity.name}SortOrder {<#
  if(entity.fields.length > 0){
  for (let key of entity.fields){
#>
  #{key}Asc
  #{key}Desc
<#-
    }
  } else {
#>
  native
  <#}#>
}