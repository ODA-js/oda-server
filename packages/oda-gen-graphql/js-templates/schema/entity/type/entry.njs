<#@ context 'entity'#>
<#@ alias 'type-entry'#>

type #{entity.name} <#-if(entity.implements.length > 0){#>
implements #{entity.implements.join(' & ')}
<#}-#>
{
<#- entity.fields.forEach(field => { -#>
<# if(field.description){#>
  # #{field.description}
<#- }#>
  #{field.name}#{field.args}: #{field.type}
<#-})-#>

<#- entity.relations.forEach(rel => { -#>
<# if(rel.description){#>
  # #{rel.description}
<#- }#>
  <#if(rel.embedded) {-#>
  #{rel.name}#{rel.args}: <#if(!rel.single){#>[<#}#>#{rel.type}<#if(!rel.single){#>]<#}#>
  <#-} else if(rel.single) {-#>

  #{rel.name}#{rel.args}: #{rel.type}

  <#-} else {-#>
  <#if(rel.derived){#>
  #{rel.name}(#{rel.args} ): #{rel.connectionName}
  <#} else {#>
  #{rel.name}(after: String, first: Int, before: String, last: Int, limit: Int, skip: Int, orderBy: [#{rel.entity}SortOrder], filter:#{rel.entity}ComplexFilter ): #{rel.connectionName}
  <#-}-#>
  <#-}-#>
<#-})#>
}