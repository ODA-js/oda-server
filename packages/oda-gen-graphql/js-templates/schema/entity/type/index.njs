<#@ chunks "$$$main$$$" -#>
<#@ context 'entity' #>
<#@ alias 'type-index'#>

<#- chunkStart(`./type/index.ts`); -#>
#{partial(entity.type.resolver, 'type-resolver-imports')}
export default new Type({
  schema: gql`
    #{partial(entity.type.entry, 'type-entry')}
  `,
  resolver: {
    id: ({ id }) => id,
    #{partial(entity.type.resolver, 'type-resolver')}
  }
});