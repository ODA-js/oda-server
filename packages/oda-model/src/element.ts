import { IValidate, IValidator, IValidationResultInput } from './validation';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';

export const Internal = Symbol.for('internal');
export const MetaData = Symbol.for('metadata');

export interface ElementMetaInfo {}

export interface ElementInternal<T extends ElementMetaInfo> {
  [MetaData]: T;
}

export interface ElementInput<T extends ElementMetaInfo> {
  metadata?: Partial<T>;
}

export interface ElementOutput<T extends ElementMetaInfo> {
  metadata: T;
}

export interface IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: any): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): any;
}

/**
 * updatable items
 */
export interface IUpdatable<
  T extends ElementMetaInfo,
  I extends ElementInput<T>,
  O extends ElementOutput<T>
> extends IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: Nullable<I>): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): O;
}

/**
 * Meta interface for updating items with metadata
 */
export interface IMeta<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  O extends ElementOutput<M>
> extends IUpdatable<M, I, O>, IValidate {
  /**
   * the kind of current item
   */
  readonly modelType: MetaModelType;
  /**
   * meta information that is outside of model notations and can be customized as well
   */
  readonly metadata: Readonly<M>;
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Element<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  S extends ElementInternal<M>,
  O extends ElementOutput<M>
> implements IMeta<M, I, O> {
  public get modelType(): MetaModelType {
    return 'element';
  }
  public [Internal]: S;
  /**
   * metadata is not immutable
   */
  public get metadata(): M {
    return this[Internal][MetaData];
  }

  public validate(validator: IValidator): IValidationResultInput[] {
    return validator.check(this);
  }

  constructor(init: ElementInput<M>) {
    this[Internal] = {} as S;
    this.updateWith(merge({}, defaultInput, init));
  }
  /**
   * patches metadata for entity
   * @param input metadata patch object
   */
  public updateWith(input: Nullable<ElementInput<M>>) {
    assignValue<Element<M, I, S, O>, ElementInput<M>, ElementInternal<M>>({
      src: this,
      input,
      field: 'metadata',
      effect: (_, value) =>
        (this[Internal][MetaData] = merge({}, this[Internal][MetaData], value)),
      setDefault: _ =>
        (this[Internal][MetaData] = merge({}, defaultMetaInfo) as M),
    });
  }
  /**
   * make item usable within another constructor
   * it is suitable for cloning items
   */
  public toObject(): O {
    return merge({}, { metadata: this.metadata } as O);
  }
}
