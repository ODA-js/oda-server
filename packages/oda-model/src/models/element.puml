@startuml

interface IMeta
interface IUpdatable
interface IValidate

IUpdatable <|-- IMeta
IValidate <|-- IMeta

class Element

IMeta <|-- Element


@enduml
