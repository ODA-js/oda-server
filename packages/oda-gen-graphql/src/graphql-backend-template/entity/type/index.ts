import * as entry from './entry.graphql';
import * as resolver from './resolver.ts';
import * as enums from './enums.graphql';

export { entry, resolver, enums };
