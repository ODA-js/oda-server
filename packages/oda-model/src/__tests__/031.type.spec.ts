import 'jest';
import { MetaData } from '../element';
import { Type, TypeDefaultMetaInfo } from '../type';

describe('Type', () => {
  it('default', () => {
    const res = new Type({ name: 'A' });
    expect(res[MetaData]).toMatchObject(TypeDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('default has no Singular', () => {
    const res = new Type({ name: 'Species' });
    expect(res.plural).toBe('AllSpecies');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('should Capitalize name', () => {
    const res = new Type({ name: 'entity' });
    expect(res.name).toBe('entity');
  });
  it('take TitlePlural from Plural name', () => {
    const res = new Type({ name: 'Child', plural: 'Children  ' });
    expect(res.plural).toBe('Children');
    expect(res.titlePlural).toBe('Children');
  });
  it('should take Plural Title plural', () => {
    const res = new Type({
      name: 'Species',
      plural: 'Species',
      titlePlural: 'All Species',
    });
    expect(res.titlePlural).toBe('All Species');
    expect(res.plural).toBe('AllSpecies');
  });

  it('creates specified field with specified types', () => {
    const res = new Type({
      name: 'A',
      fields: [
        {
          name: 'A',
        },
        {
          name: 'B',
          type: { type: 'enum', name: 'BBB' },
        },
        {
          name: 'c',
          type: { type: 'entity', name: 'CCC', multiplicity: 'many' },
        },
      ],
    });
    expect(res.fields.size).toBe(3);
    expect(res.fields.get('A')).not.toBeUndefined();
    const a = res.fields.get('a');
    if (a) {
      expect(a.modelType).toBe('input-simple-field');
    }
    expect(res.fields.get('A')).not.toBeUndefined();
    const b = res.fields.get('B');
    if (b) {
      expect(b.modelType).toBe('input-enum-field');
    }
    const c = res.fields.get('c');
    if (c) {
      expect(c.multiplicity).toBe('many');
      expect(c.modelType).toBe('input-entity-field');
    }
    expect(res).toMatchSnapshot('object with fields');
    expect(res.toObject()).toMatchSnapshot('toObject with fields');
  });

  it('should pass entity name to fields', () => {
    const res = new Type({
      name: 'A',
      fields: [{ name: 'a' }],
    });
    const a = res.fields.get('a');
    if (a) {
      expect(a.metadata.entity).toBe('A');
      expect(a.metadata.order).toBe(0);
    }
  });

  it('should deduplicates fields', () => {
    const res = new Type({
      name: 'A',
      fields: [
        { name: 'a' },
        { name: 'a', description: 'field a description' },
      ],
    });
    expect(res.fields.size).toBe(1);
  });
});
