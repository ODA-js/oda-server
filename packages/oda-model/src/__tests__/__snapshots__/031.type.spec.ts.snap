// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type creates specified field with specified types: object with fields 1`] = `
Record {
  Symbol(internal): Object {
    "fields": Map {
      "A" => RecordField {
        Symbol(internal): Object {
          "kind": "input",
          "multiplicity": "one",
          "name": "A",
          "type": "string",
        },
        Symbol(metadata): Object {
          "order": 0,
          "required": false,
        },
      },
      "B" => RecordField {
        Symbol(internal): Object {
          "kind": "input",
          "multiplicity": "one",
          "name": "B",
          "type": Object {
            "multiplicity": "one",
            "name": "BBB",
            "type": "enum",
          },
        },
        Symbol(metadata): Object {
          "order": 1,
          "required": false,
        },
      },
      "c" => RecordField {
        Symbol(internal): Object {
          "kind": "input",
          "multiplicity": "many",
          "name": "c",
          "type": Object {
            "multiplicity": "many",
            "name": "CCC",
            "type": "entity",
          },
        },
        Symbol(metadata): Object {
          "order": 2,
          "required": false,
        },
      },
    },
    "kind": "input",
    "name": "A",
  },
  Symbol(metadata): Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "global": "A",
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
}
`;

exports[`Type creates specified field with specified types: toObject with fields 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {
        "order": 0,
        "required": false,
      },
      "multiplicity": "one",
      "name": "A",
      "title": undefined,
      "type": "string",
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {
        "order": 1,
        "required": false,
      },
      "multiplicity": "one",
      "name": "B",
      "title": undefined,
      "type": Object {
        "multiplicity": "one",
        "name": "BBB",
        "type": "enum",
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {
        "order": 2,
        "required": false,
      },
      "multiplicity": "many",
      "name": "c",
      "title": undefined,
      "type": Object {
        "multiplicity": "many",
        "name": "CCC",
        "type": "entity",
      },
    },
  ],
  "kind": "input",
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "global": "A",
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
  "name": "A",
  "title": undefined,
}
`;

exports[`Type default has no Singular: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [],
  "kind": "input",
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "global": "Species",
    "name": Object {
      "plural": "AllSpecies",
      "singular": "Species",
    },
    "titlePlural": "AllSpecies",
  },
  "name": "Species",
  "title": undefined,
}
`;

exports[`Type default: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [],
  "kind": "input",
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "global": "A",
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
  "name": "A",
  "title": undefined,
}
`;
